inference_server/model_handler/deployment.py

```diff
@@ -0,0 +1,131 @@
+
+    def dict_to_proto(self, generate_kwargs: dict) -> dict:
+        result = {}
+        for k, v in generate_kwargs.items():
+            if v is not None:
+                x = generation_pb2.Value()
+                setattr(x, self.dtype_proto_field[type(v)], v)
+                result[k] = x
+
+        return result
+
+    def _initialize_service(self, args: argparse.Namespace):
+        if self._is_socket_open(self.port_number):
+            raise RuntimeError(
+                f"Server is already running on port {self.port_number}, please shutdown or use different port."
+            )
+
+        cmd = f"inference_server.model_handler.launch --model_name {args.model_name} --deployment_framework {args.deployment_framework} --dtype {get_str_dtype(args.dtype)} --port {self.port_number}"
+
+        if args.deployment_framework in [DS_INFERENCE, DS_ZERO]:
+            cmd = f"deepspeed --num_gpus {self.num_gpus} --module {cmd}"
+        else:
+            raise NotImplementedError(f"unsupported deployment_framework: {args.deployment_framework}")
+
+        cmd = cmd.split(" ")
+        self.process = subprocess.Popen(cmd)
+
+    def _initialize_grpc_client(self):
+        channels = []
+        self.stubs = []
+        for i in range(self.num_gpus):
+            channel = grpc.aio.insecure_channel(f"localhost:{self.port_number + i}")
+            stub = generation_pb2_grpc.GenerationServiceStub(channel)
+            channels.append(channel)
+            self.stubs.append(stub)
+
+    # runs task in parallel and return the result from the first task
+    async def _query_in_tensor_parallel(self, text: List[str], generate_kwargs: dict):
+        responses = []
+        for i in range(self.num_gpus):
+            responses.append(self.asyncio_loop.create_task(self._request_async_response(i, text, generate_kwargs)))
+
+        await responses[0]
+        return responses[0]
+
+    async def _request_async_response(self, stub_id: int, text: List[str], generate_kwargs: dict):
+        req = generation_pb2.GenerationRequest(texts=text, generate_kwargs=generate_kwargs)
+        response = await self.stubs[stub_id].Generate(req)
+        return response
+
+    def generate(self, **kwargs) -> GenerateResponse:
+        if self.use_grpc_server:
+            if "request" in kwargs:
+                request = kwargs["request"]
+                text = request.text
+                generate_kwargs = request.get_generate_kwargs()
+                generate_kwargs = self.dict_to_proto(generate_kwargs)
+            else:
+                text = kwargs["text"]
+                generate_kwargs = self.dict_to_proto(kwargs["generate_kwargs"])
+
+            response = self.asyncio_loop.run_until_complete(
+                self._query_in_tensor_parallel(text, generate_kwargs)
+            ).result()
+
+            response = GenerateResponse(
+                text=[r for r in response.texts], num_generated_tokens=[n for n in response.num_generated_tokens]
```

Comment:
nit: I don't think the comprehensions are needed?

Reply:
Yes, they are. response.texts is return as some GRPC class and there is no list() method for them.